src/util/store/conflict-resolution.ts

```diff
@@ -68,60 +71,97 @@ export function resolvedDocumentUpdates(
             'Unexpected error: missing left document in conflict resolution'
           )
 
-        const leftFileDoc = asMaybe(asStoreFileDocument)(leftDoc)
-        const rightFileDoc = asMaybe(asStoreFileDocument)(rightDoc)
-
-        // Merge file documents
-        if (leftFileDoc != null && rightFileDoc != null) {
-          const [winningDoc, losingDoc] = sortStoreFileDocuments(
-            leftFileDoc,
-            rightFileDoc
-          )
-          return [winningDoc, losingDoc, ...sortedDocuments]
-        }
-
-        // Assert that there should be no deleted documents while resolving conflicts
-        if ('_deleted' in leftDoc || '_deleted' in rightDoc) {
-          throw new Error(`Unexpected deleted document in conflict resolution`)
-        }
-
-        // Unknown document types
-        throw new Error('Unexpected document types in conflict resolver')
+        return trial<StoreDocument[]>(
```

Comment:
I'm not sure if this abstraction is helpful. I *think* what you are trying to do is merge files (if they are files) or merge repo documents (if they are repo documents).

Wouldn't it be easier to do something like this:

```
const leftFile = asMaybe(asStoreFileDocument)(leftDoc)
const rightFile = asMaybe(asStoreFileDocument)(rightDoc)
if (leftFile != null && rightFile != null) return mergeFiles(leftFile, rightFile)

const leftRepo = asMaybe(asStoreRepoDocument)(leftDoc)
const rightRepo = asMaybe(asStoreRepoDocument)(rightDoc)
if (leftRepo != null && rightRepo != null) return mergeRepos(leftRepo, rightRepo)

throw new TypeError('Expecting matching document types')
```

Sure, that final `throw` doesn't capture all the information it possibly could, but I think that's fine for code that should never actually run (unless we have other horrible problems with the database or server).

Of course, I am open to being persuaded the other way. Sometimes abstractions are necessary if there are hard problems to solve, but this just doesn't seem like one of those at first glance.

Reply:
I am approving anyhow, since I do believe the code works as-is.

Reply:
Capturing all the errors _is_ the motivation for this abstraction. My persuasive argument here is that for code that should never run, I'd like as much information as possible about arriving at this failure case to expedite resolving the error.